### PR TITLE
#3180. Add tests for `FunctionToJSExportedDartFunction`. Part 2.

### PR DESCRIPTION
**Pull request by sgrekhov**
Please note that I use `Expect.equals(globalContext, pThis);` to check the value of `this`. Is it Ok to make an assumption that `this == globalContext`?

### CODE REVIEW

**Review by eernstg — COMMENTED** [review]
Added a few comments. The main thing is that I'm surprised that `toJSCaptureThis` is documented as being identical to `toJS` except for capturing and passing `this`, but the requirements on the types used in the signature of the receiver seem to be different (it looks like nothing is required for `pThis`).

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t03.dart [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+/// A callable JavaScript function that wraps this [Function] and captures the
+/// `this` value when called.
+///
+/// Identical to `toJS`, except the resulting [JSExportedDartFunction] will pass
+/// `this` from JavaScript as the first argument to the converted [Function].
+/// Any [Function] that is converted with this member should take in an extra
+/// parameter at the beginning of the parameter list to handle this.
+///
+/// @description Check that if this getter is called on a torn off constructor
+/// it is a run-time error to call it in JavaScript.
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t03.dart]: Yes, thank you. Forgot to updated the description. Fixed.

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t01.dart [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+/// @assertion JSExportedDartFunction get toJSCaptureThis
+/// A callable JavaScript function that wraps this [Function] and captures the
+/// `this` value when called.
+///
+/// Identical to `toJS`, except the resulting [JSExportedDartFunction] will pass
+/// `this` from JavaScript as the first argument to the converted [Function].
+/// Any [Function] that is converted with this member should take in an extra
+/// parameter at the beginning of the parameter list to handle this.
+///
+/// @description Check that this property returns JavaScript function that wraps
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t01.dart]: Fixed everywhere.

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t01.dart [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+/// this [Function] and captures the `this` value when called.
+/// @author sgrekhov22@gmail.com
+
+import 'dart:js_interop';
+import 'dart:js_interop_unsafe';
+import '../../../Utils/expect.dart';
+import '../js_utils.dart';
+
+int foo(pThis) {
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t01.dart:23]: Awaiting for the clarification.

**Review by sgrekhov — COMMENTED** [review]
Updated but waitig for further info.

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t01.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (srujzs) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t01.dart:23]: Sergey filed a few issues, but the central problem is that I never duplicated the static error checks for `toJSCaptureThis`: https://github.com/dart-lang/sdk/issues/61439.

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t01.dart [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  Expect.equals(globalContext, pThis);
```
[the remark this answers is not in the document]
Comment (srujzs) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t01.dart:24]: Once `pThis` is updated to e.g. `JSAny`, the warnings around equality and identicality are the same as other instances of JS interop. Equality will check the underlying JS values, but identicality will check the wrappers and therefore fail in dart2wasm.

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t02.dart [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+class C {
+  A a;
+  C(): a = A(42);
+
+  int foo(JSObject pThis) {
+    Expect.equals(globalContext, pThis);
+    return a.id;
+  }
+  static String bar(pThis, String v) {
+    Expect.equals(globalContext, pThis);
```
[the remark this answers is not in the document]
Comment (srujzs) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t02.dart:38]: Another alternative is using `callAsFunction` (or the JS equivalent: `call`) which allows you to pass a `this` value explicitly.

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t03.dart [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import '../../../Utils/expect.dart';
+import '../js_utils.dart';
+
+@JS()
+extension type ET(JSNumber id) {
+  ET.c1(pThis, this.id) {
+    Expect.equals(globalContext, pThis);
+  }
+
+  @JS()
```
Comment (srujzs) [inline LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t03.dart:29]: Note that `@JS` does nothing on constructors here and below, regardless of `external` or not. It's still on my plate to add warnings though.

**Review by srujzs — APPROVED** [review]

**Review by eernstg — COMMENTED** [review]
Very good, I think this one is now subject to a few changes, and there may be another PR with some other changes. So I'll just await further input on this PR for now.

**Comment by sgrekhov** [comment]
Thank you! Updated.
- Removed `@JS()` from the constructors (otherwise the tests fail once these annotations start producing warnings).
- Changed the type of the `pThis` parameter to `JSObject` (again, otherwise the tests will fail when type checking is added). Please let me know if this type should instead be `JSAny`.
- Added a test checking arguments of types `JSObject` and `JSAny`.

Please take another look.

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t02.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (eernstg) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t02.dart:38]: @sgrekhov, I think this can be settled in a separate PR.

[the remark this answers is not in the document]
Comment (sgrekhov) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t02.dart:38]: Yes, I'll add these tests in a separate PR

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A02_t02.dart [hunk 7]
```diff
@@ -20,15 +20,17 @@ import 'dart:js_interop';
 
 class C {}
 
-C? foo(pThis) {return null;}
+C? foo(JSObject pThis) {return null;}
 
-void bar(pThis, C c) {}
+void bar(JSObject pThis, C c) {}
```
Comment (srujzs) [inline LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A02_t02.dart:25]: nit: It may be worth having a function with all valid return and parameter types except for the first parameter (the `this` value) to check that we still catch that.

**Review by srujzs — APPROVED** [review]

**Comment by srujzs** [comment]
> Please let me know if this type should instead be JSAny.

It can be any valid type for interop members. `JSObject` or `JSAny` works in this case because `globalThis` is a JS object, but if we add tests to pass a symbol, `null`, etc. for `this`, then we'd naturally want to change the type we write.

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A02_t02.dart (hunk 7 above)
Comment (sgrekhov) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A02_t02.dart:25]: There is `void corge(pThis) {}` on the line 33.

Comment (srujzs) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A02_t02.dart:25]: Ah I missed that, thanks!

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t04.dart [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+import '../../../Utils/expect.dart';
+import '../js_utils.dart';
+
+String foo(JSObject pThis, String s, bool b, num n, int i, double d) {
+  Expect.equals(globalContext, pThis);
+  return "foo($s, $b, $n, $i, $d)";
+}
+
+JSString bar(
+  pThis,
```
Comment (eernstg) [inline LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t04.dart:29]: Probably:
```suggestion
  JSObject pThis,
```

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A02_t02.dart [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+void bar(JSObject pThis, C c) {}
+
+void baz(JSObject pThis, Map<String, int> s) {}
+
+dynamic qux(JSObject pThis) {}
+
+void quux(JSObject pThis, Symbol s) {}
+
+void corge(pThis) {}
```
Comment (eernstg) [inline LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A02_t02.dart:33]: (This one was already mentioned somewhere)
```suggestion
void corge(JSObject pThis) {}
```

**Review by eernstg — APPROVED** [review]
Looks good, but it looks like one parameter type is still missing in a couple of places. Otherwise I think it's ready to land.

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t02.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t02.dart:38]: > Another alternative is using `callAsFunction` (or the JS equivalent: call) which allows you to pass a this value explicitly.

Ups! There are no tests for `callAsFunction`. I'll add them too.

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A02_t02.dart (hunk 9 above)
Comment (eernstg) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A02_t02.dart:33]: Oh, of course, there error may seem to be here, but the error message is at the invocation of `toJSCaptureThis`.

File: LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t02.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t02.dart:38]: > Another alternative is using `callAsFunction` (or the JS equivalent: `call`) which allows you to pass a `this` value explicitly.

@srujzs I'm trying to check value of `this` via `callAsFunction`. The code below works on `dart2was` but in case of `dart2js` it produces `TypeError: Instance of 'String': type 'UnknownJavaScriptObject' is not a subtype of type 'String'`. Am I doing anything wronr? Or this is an issue?

```dart
void foo(JSObject pThis) {
  Expect.equals("Value of this", (pThis as JSString).toDart); // Error here on dart2js
}

main() {
  globalContext["jsFoo"] = foo.toJSCaptureThis;
  JSFunction jsFoo = globalContext["jsFoo"] as JSFunction;
  jsFoo.callAsFunction("Value of this".toJS);
}
```

[the remark this answers is not in the document]
Comment (srujzs) [reply LibTest/js_interop/FunctionToJSExportedDartFunction/toJSCaptureThis_A01_t02.dart:38]: That's because `pThis` isn't a `JSObject` under the hood :D

You likely want `void foo(String pThis)`. It works in dart2wasm because `JSObject` and `JSString` are the same box. In dart2js and DDC, we have stronger type-checking.
